### PR TITLE
Set default RMM pool to ASYNC for cuda 11.2+

### LINKED ISSUE #4515: [FEA] Set RMM async allocator as default
**Is your feature request related to a problem? Please describe.**
Currently the RMM arena allocator is the default, but in some circumstances it can still run into OOM errors due to memory fragmentation. The async allocator, relying on `cudaMallocAsync` and `cudaFreeAsync`, can remap physical pages when running of GPU memory, thus is more resistant to memory fragmentation.

**Describe the solution you'd like**
Set the async allocator as the default.

**Describe alternatives you've considered**
Continue improving the arena allocator is possible, but since it doesn't remap memory, probably won't ever be as fragmentation resistance as the async allocator.

**Additional context**
Set this at the beginning of 22.04 for more testing.

### PR DESCRIPTION
**Pull request by rongou**
Fixes #4515 

Signed-off-by: Rong Ou <rong.ou@gmail.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -404,10 +406,11 @@ object RapidsConf {
...
       "\"ARENA\", the RMM arena allocator is used; with \"ASYNC\", the new CUDA stream-ordered " +
       "memory allocator in CUDA 11.2+ is used. If set to \"NONE\", pooling is disabled and RMM " +
-      "just passes through to CUDA memory allocation directly. Note: \"ARENA\" is the " +
-      "recommended pool allocator if CUDF is built with Per-Thread Default Stream (PTDS).")
+      "just passes through to CUDA memory allocation directly. When unset, if the CUDA driver is " +
+      "newer than 11.2, the \"ASYNC\" allocator is used; otherwise, the \"ARENA\" allocator is " +
+      "used.")
     .stringConf
-    .createWithDefault("ARENA")
+    .createWithDefault("")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:413]: Given the number of users on CUDA < 11.2 will be dwindling over time quickly, wondering if we should just have `ASYNC` automatically fallback to `ARENA` with a log message on startup when the fallback occurs.

**Review by jlowe — COMMENTED** [review]
Are we really ready for this?  It would be good to have benchmark results showing this is not going to be a regression for some use-cases.  It seems like we still have unresolved performance issues that appear to be tied to this allocator, e.g.:#4536.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 1 above)
Comment (rongou) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: Done.

**Comment by rongou** [comment]
We do have benchmarks on TPCDS at various scale factors. The idea is turn this on early in 22.04 and get it more widely tested and benchmarked. If we do run into some block we can always revert it back before the release is cut.

**Comment by jlowe** [comment]
> We do have benchmarks on TPCDS at various scale factors.

Do we have those metrics isolated to just this change?  There are quite a few performance improvements going in recently, and I want to make sure we don't end up masking a regression from the async allocator with a performance improvement from elsewhere.

I would really like to see what this allocator costs for queries that don't spill, and last I knew we only had metrics at scale for UCX shuffle which of course can spill quite a lot due to its aggressive caching of shuffle outputs.  The only metrics I've seen for a query that doesn't spill is from #4536, and I find it odd that we want to check this in without first understanding the impact footprint for a 500X slowdown.

**Comment by revans2** [comment]
> I would really like to see what this allocator costs for queries that don't spill, and last I knew we only had metrics at scale for UCX shuffle which of course can spill quite a lot due to its aggressive caching of shuffle outputs. The only metrics I've seen for a query that doesn't spill is from #4536, and I find it odd that we want to check this in without first understanding the impact footprint for a 500X slowdown.

To be clear most of the tests didn't show that slowness. I would need to do some more analysis to see if there was any real difference that I saw beyond this one corner case.

**Comment by rongou** [comment]
This is the result for TPCDS at 1TB. I'll dig more into #4536.
ARENA vs ASYNC allocator
.

**Comment by rongou** [comment]
#4536 turns out to be a red herring.

**Review by jlowe — APPROVED** [review]

**Comment by rongou** [comment]
build

**Review by abellina — APPROVED** [review]

**Pull request merged**